src/app/InteractionModelEngine.cpp

```diff
@@ -764,34 +867,132 @@ CHIP_ERROR InteractionModelEngine::CanEstablishReadTransaction(const ReadHandler
+        usedReadHandlersInFabric++;
         return Loop::Continue;
     });
 
-    // The incoming read handler here is also counted above.
-    if (activeReadHandlersOnCurrentFabric > kReservedReadHandlersPerFabricForReadRequests)
+    // Busy, since there is already some read requests ongoing on this fabric, please retry later.
+    if (usedAttributePathsInFabric + aRequestedAttributePathCount >
+            kMinSupportedPathsPerReadRequest * guaranteedReadRequestsPerFabric ||
+        usedEventPathsInFabric + aRequestedEventPathCount > kMinSupportedPathsPerReadRequest * guaranteedReadRequestsPerFabric ||
+        usedReadHandlersInFabric >= guaranteedReadRequestsPerFabric)
     {
-        return (allowUnlimited ? CHIP_NO_ERROR : CHIP_IM_GLOBAL_STATUS(Busy));
+        return Protocols::InteractionModel::Status::Busy;
     }
 
-    return CHIP_NO_ERROR;
+    const auto evictAndUpdateResourceUsage = [&](FabricIndex fabricIndex) {
+        bool ret = TrimFabricForRead(fabricIndex);
+        countResourceUsage();
+        return ret;
+    };
+
+    //
+    // At this point, we have an inbound request that respects minimas but we still don't have enough resources to handle it. Which
+    // means that we definitely have handlers on existing fabrics that are over limits and need to evict at least one of them to
+    // make space.
+    //
+    bool didEvictHandler = true;
+    while (didEvictHandler)
+    {
+        didEvictHandler = false;
+        didEvictHandler = didEvictHandler || evictAndUpdateResourceUsage(kUndefinedFabricIndex);
+        if (usedAttributePaths + aRequestedAttributePathCount <= attributePathCap &&
+            usedEventPaths + aRequestedEventPathCount <= eventPathCap && usedReadHandlers < readHandlerCap)
+        {
+            break;
+        }
+        // If the fabric table is full, we won't evict read requests from normal fabrics before we have evicted all read requests
+        // from PASE sessions.
+        if (mpFabricTable->FabricCount() == GetConfigMaxFabrics() && didEvictHandler)
+        {
+            continue;
+        }
+        for (const auto & fabric : *mpFabricTable)
+        {
+            // The resources are enough to serve this request, do not evict anything.
+            if (usedAttributePaths + aRequestedAttributePathCount <= attributePathCap &&
+                usedEventPaths + aRequestedEventPathCount <= eventPathCap && usedReadHandlers < readHandlerCap)
+            {
+                break;
+            }
+            didEvictHandler = didEvictHandler || evictAndUpdateResourceUsage(fabric.GetFabricIndex());
+        }
+    }
+
+    // The read requests that are not using exceeded resources will be first-come-first-served, since the read transactions will be
+    // relatively short, the client is expected to retry later if the server is busy now.
+    if (usedAttributePaths + aRequestedAttributePathCount > attributePathCap ||
+        usedEventPaths + aRequestedEventPathCount > eventPathCap)
```

Comment:
How can we end up in this situation?  I would have thought that at this point we guarantee that we have enough resources for the incoming read, since our current fabric was not using its max quota of read handlers or paths, which means some other fabric was over quota, which means we should have evicted something and gotten ourselves enough free space for this, no?

Reply:
Yes, this should not happen, since all fabric (after accepting this one) will not be oversized.